crates/p2p/src/swarm/mod.rs

```diff
@@ -536,58 +548,249 @@ where
         Ok(())
     }
 
-    #[instrument(skip(self, _event))]
-    async fn handle_response_request_event(
+    #[instrument(skip(self, event))]
+    async fn handle_request_response_event(
+        &mut self,
+        event: RequestResponseEvent<GetMessageRequest, GetMessageResponse>,
+    ) -> P2PResult<()> {
+        let RequestResponseEvent::Message { peer, message } = event else {
+            return Ok(());
+        };
+
+        match message {
+            request_response::Message::Request {
+                request_id,
+                request,
+                channel,
+            } => {
+                let Some(req) = v1::GetMessageRequest::from_msg(request) else {
+                    debug!(%peer, "Peer sent invalid get message request, disconnecting it");
+                    let _ = self.swarm.disconnect_peer_id(peer);
+                    return Ok(());
+                };
+
+                let msg = self.handle_get_message_request(req).await?;
+
+                if msg.is_none() {
+                    debug!(%request_id, "Have no needed data, requesting from neighbours");
+                    return Ok(()); // TODO(NikitaMasych): launch recursive request.
+                }
+
+                let response = GetMessageResponse {
+                    msg: vec![msg.unwrap()],
+                };
+
+                let _ = self
+                    .swarm
+                    .behaviour_mut()
+                    .request_response
+                    .send_response(channel, response)
+                    .inspect_err(|_| debug!("Failed to send response"));
+            }
+
+            request_response::Message::Response {
+                request_id,
+                response,
+            } => {
+                if response.msg.is_empty() {
+                    debug!(%request_id, "Have no needed data, requesting from neighbours");
+                    return Ok(()); // TODO(NikitaMasych): launch recursive request.
+                }
+                let mut all_messages_empty = true;
+                for msg in response.msg.into_iter() {
+                    if msg.body.is_none() {
+                        continue;
+                    }
+
+                    // TODO: report/punish peer for invalid message?
+                    match GossipsubMsg::from_proto(msg.clone()) {
+                        Ok(msg) => {
+                            if let Err(err) = validate_gossipsub_msg::<DSP>(peer, &msg) {
+                                debug!(%peer, reason=%err, "Got invalid signature from peer, rejecting message.");
+                                continue;
+                            }
+                        }
+                        Err(err) => {
+                            debug!(%peer, reason=%err, "Peer sent invalid message");
+                            continue;
+                        }
+                    }
```

Comment:
To reduce nesting move validation from match branch 

```suggestion
                    let msg = match GossipsubMsg::from_proto(msg.clone()) {
                        Ok(msg) => {
                            msg
                        }
                        Err(err) => {
                            debug!(%peer, reason=%err, "Peer sent invalid message");
                            continue;
                        }
                    }
                    
                    if let Err(err) = validate_gossipsub_msg::<DSP>(peer, &msg) {
                        debug!(%peer, reason=%err, "Got invalid signature from peer, rejecting message.");
                        continue;
                    }
```

Reply:
Basically, rationale for calling `from_proto` is only in what we would like to use `validate_gossipsub_msg` method; that hints what i would like to put it in one code block. We can use arm guard tho:

```rust
 match GossipsubMsg::from_proto(msg.clone()) {
    Ok(msg) if validate_gossipsub_msg::<DSP>(peer, &msg).is_err() => {
        debug!(%peer, "Got invalid signature from peer, rejecting message.");
        continue;
    }
    Err(err) => {
        debug!(%peer, %err, "Peer sent invalid message");
        continue;
    }
    _ => {}
}
```

Because this part in suggested snippet looks a bit pointless tbh:

```rust
Ok(msg) => {
    msg
}
```

Reply:
Ah, nvm, I've checked your comments below about using typed `GossipsubMsg` not only here

Reply:
And that's okay, as you need it later in `handle_get_message_response` I suggest getting `msg` from it to not decode the "raw" data again in `handle_get_message_response` as you did

Reply:
Also, this one is not great too, as you omit the error

```rust
    Ok(msg) if validate_gossipsub_msg::<DSP>(peer, &msg).is_err() => {
```